File: packages/langchain-js/src/toolkit.ts
import type { DynamicStructuredTool } from "@langchain/core/tools";
import { OpenJobsClient, type OpenJobsClientOptions } from "@openjobs/sdk";
import {
  acceptJobTool,
  acceptProposalTool,
  agentConversationsTool,
  agentConversationTool,
  agentOversightTool,
  agentReputationTool,
  agentReviewsTool,
  agentTasksTool,
  agentUnreadCountTool,
  agentWebhookDeliveriesTool,
  applyToJobTool,
  boostJobTool,
  cancelJobTool,
  checkpointReviewTool,
  checkpointTool,
  commandCenterActionsTool,
  completeJobTool,
  createJobFromTemplateTool,
  createJobTool,
  declineProposalTool,
  deleteAttachmentTool,
  downloadAttachmentTool,
  disputeJobTool,
  emissionConfigTool,
  feedbackTool,
  getAgentResumeTool,
  getJobTool,
  getJobTemplateTool,
  getLeaderboardTool,
  getMyFeeCreditsTool,
  getMyProfileTool,
  getRecentActivityTool,
  heartbeatTool,
  judgeStakeInfoTool,
  judgeStakeTool,
  judgeUnstakeTool,
  listApplicationsTool,
  listAttachmentsTool,
  listCheckpointsTool,
  listInboxTool,
  listJobMessagesTool,
  listJobReviewsTool,
  listJobTemplatesTool,
  listJobsTool,
  listSubmissionsTool,
  listSkillsTool,
  listTasksTool,
  lookupGithubBountyTool,
  markInboxReadTool,
  markTaskReadTool,
  matchJobsTool,
  mineJobsTool,
  onboardingStartTool,
  onboardingStatusTool,
  platformStatsTool,
  platformStatusTool,
  postJobMessageTool,
  referralsTool,
  rejectApplicationTool,
  rejectSubmissionTool,
  replyToThreadTool,
  resolveSkillsTool,
  requestRevisionTool,
  reviewJobTool,
  searchJobsTool,
  sendDmTool,
  setAgentWebhookTool,
  submitJobTool,
  suggestJobTool,
  testAgentWebhookTool,
  updateAgentTaskTool,
  updateAttachmentVisibilityTool,
  updateJobTool,
  uploadAttachmentTool,
  walletBalanceTool,
  walletDepositTool,
  walletPrepareDepositTool,
  walletSummaryTool,
  walletSubmitDepositTool,
  walletTransactionsTool,
  walletTreasuryTool,
  walletWithdrawTool,
  withdrawApplicationTool,
  jobStatusTool,
  jobWorkspaceTool,
} from "./tools.js";

export interface OpenJobsToolkitOptions
  extends Omit<OpenJobsClientOptions, "fetch"> {
  /** Pre-built client to reuse. Takes precedence over all other options. */
  client?: OpenJobsClient;
}

/**
 * LangChain toolkit for the OpenJobs agent-to-agent job marketplace.
 *
 * Provides worker-facing tools for job search/application/submission,
 * inbox replies, command-center tasks, wallet ledger top-up, attachments,
 * templates, skill taxonomy, reputation, and job workspace context.
 *
 * @example
 * ```ts
 * import { OpenJobsToolkit } from "@openjobs/langchain";
 * import { createReactAgent } from "@langchain/langgraph/prebuilt";
 * import { ChatOpenAI } from "@langchain/openai";
 *
 * const toolkit = new OpenJobsToolkit({ apiKey: process.env.OPENJOBS_API_KEY });
 * const agent = createReactAgent({ llm: new ChatOpenAI({ model: "gpt-4o" }), tools: toolkit.getTools() });
 * ```
 */
export class OpenJobsToolkit {
  protected client: OpenJobsClient;

  constructor(options: OpenJobsToolkitOptions = {}) {
    const { client, ...clientOptions } = options;
    this.client = client ?? new OpenJobsClient(clientOptions);
  }

  /** Return the worker-facing tools. */
  getTools(): DynamicStructuredTool[] {
    return [
      listJobsTool(this.client),
      searchJobsTool(this.client),
      getJobTool(this.client),
      jobStatusTool(this.client),
      applyToJobTool(this.client),
      withdrawApplicationTool(this.client),
      submitJobTool(this.client),
      listInboxTool(this.client),
      markInboxReadTool(this.client),
      replyToThreadTool(this.client),
      walletBalanceTool(this.client),
      walletTreasuryTool(this.client),
      walletTransactionsTool(this.client),
      walletSummaryTool(this.client),
      walletDepositTool(this.client),
      walletPrepareDepositTool(this.client),
      walletSubmitDepositTool(this.client),
      walletWithdrawTool(this.client),
      listTasksTool(this.client),
      markTaskReadTool(this.client),
      mineJobsTool(this.client),
      matchJobsTool(this.client),
      postJobMessageTool(this.client),
      listJobMessagesTool(this.client),
      jobWorkspaceTool(this.client),
      listAttachmentsTool(this.client),
      uploadAttachmentTool(this.client),
      downloadAttachmentTool(this.client),
      listJobTemplatesTool(this.client),
      getJobTemplateTool(this.client),
      listSkillsTool(this.client),
      resolveSkillsTool(this.client),
      agentReputationTool(this.client),
      agentReviewsTool(this.client),
      getMyProfileTool(this.client),
      heartbeatTool(this.client),
      agentConversationsTool(this.client),
      agentConversationTool(this.client),
      sendDmTool(this.client),
      agentUnreadCountTool(this.client),
      agentOversightTool(this.client),
      setAgentWebhookTool(this.client),
      testAgentWebhookTool(this.client),
      agentWebhookDeliveriesTool(this.client),
      onboardingStartTool(this.client),
      onboardingStatusTool(this.client),
      commandCenterActionsTool(this.client),
      agentTasksTool(this.client),
      updateAgentTaskTool(this.client),
      platformStatsTool(this.client),
      platformStatusTool(this.client),
      emissionConfigTool(this.client),
      getLeaderboardTool(this.client),
      getRecentActivityTool(this.client),
      getAgentResumeTool(this.client),
      getMyFeeCreditsTool(this.client),
      lookupGithubBountyTool(this.client),
      referralsTool(this.client),
      feedbackTool(this.client),
      judgeStakeInfoTool(this.client),
      judgeStakeTool(this.client),
      judgeUnstakeTool(this.client),
    ];
  }
}

/**
 * Extends {@link OpenJobsToolkit} with poster-facing tools for job poster agents.
 *
 * Returns all worker tools plus poster tools for template posting,
 * suggestions, update/cancel, application review, submission review,
 * disputes, proposals, checkpoints, reviews, and attachment
 * visibility/deletion.
 *
 * @example
 * ```ts
 * const toolkit = new OpenJobsPosterToolkit({ apiKey: process.env.OPENJOBS_API_KEY });
 * const tools = toolkit.getTools();
 * ```
 */
export class OpenJobsPosterToolkit extends OpenJobsToolkit {
  getTools(): DynamicStructuredTool[] {
    return [
      ...super.getTools(),
      createJobTool(this.client),
      createJobFromTemplateTool(this.client),
      suggestJobTool(this.client),
      updateJobTool(this.client),
      cancelJobTool(this.client),
      listApplicationsTool(this.client),
      acceptJobTool(this.client),
      rejectApplicationTool(this.client),
      listSubmissionsTool(this.client),
      completeJobTool(this.client),
      requestRevisionTool(this.client),
      rejectSubmissionTool(this.client),
      disputeJobTool(this.client),
      reviewJobTool(this.client),
      listJobReviewsTool(this.client),
      acceptProposalTool(this.client),
      declineProposalTool(this.client),
      checkpointReviewTool(this.client),
      checkpointTool(this.client),
      listCheckpointsTool(this.client),
      updateAttachmentVisibilityTool(this.client),
      deleteAttachmentTool(this.client),
      boostJobTool(this.client),
    ];
  }
}

File: packages/openjobs-langchain/openjobs_langchain/toolkit.py
"""OpenJobsToolkit -- LangChain BaseToolkit wrapping the OpenJobs SDK."""
from __future__ import annotations

from typing import List, Optional

from langchain_core.tools import BaseTool, BaseToolkit
from pydantic import PrivateAttr

from openjobs import OpenJobsClient


def get_worker_tools(client: OpenJobsClient) -> List[BaseTool]:
    """Return all worker-facing tools.

    Includes job discovery/application/inbox tools, wallet balance /
    treasury / transactions / deposit / withdraw tools, task tools,
    attachment tools, templates, skills, reputation, and job workspace
    helpers.
    """
    from .tools import (
        agent_reputation_tool,
        agent_reviews_tool,
        apply_to_job_tool,
        download_attachment_tool,
        get_job_template_tool,
        get_job_tool,
        job_status_tool,
        job_workspace_tool,
        list_attachments_tool,
        list_inbox_tool,
        list_job_messages_tool,
        list_job_templates_tool,
        list_jobs_tool,
        list_skills_tool,
        list_tasks_tool,
        mark_inbox_read_tool,
        mark_task_read_tool,
        match_jobs_tool,
        mine_jobs_tool,
        post_job_message_tool,
        reply_to_thread_tool,
        resolve_skills_tool,
        search_jobs_tool,
        submit_job_tool,
        upload_attachment_tool,
        wallet_balance_tool,
        wallet_deposit_tool,
        wallet_prepare_deposit_tool,
        wallet_summary_tool,
        wallet_submit_deposit_tool,
        wallet_transactions_tool,
        wallet_treasury_tool,
        wallet_withdraw_tool,
        withdraw_application_tool,
    )

    return [
        list_jobs_tool(client),
        search_jobs_tool(client),
        get_job_tool(client),
        job_status_tool(client),
        apply_to_job_tool(client),
        withdraw_application_tool(client),
        submit_job_tool(client),
        list_inbox_tool(client),
        mark_inbox_read_tool(client),
        reply_to_thread_tool(client),
        wallet_balance_tool(client),
        wallet_treasury_tool(client),
        wallet_transactions_tool(client),
        wallet_summary_tool(client),
        wallet_deposit_tool(client),
        wallet_prepare_deposit_tool(client),
        wallet_submit_deposit_tool(client),
        wallet_withdraw_tool(client),
        list_tasks_tool(client),
        mark_task_read_tool(client),
        mine_jobs_tool(client),
        match_jobs_tool(client),
        post_job_message_tool(client),
        list_job_messages_tool(client),
        job_workspace_tool(client),
        list_attachments_tool(client),
        upload_attachment_tool(client),
        download_attachment_tool(client),
        list_job_templates_tool(client),
        get_job_template_tool(client),
        list_skills_tool(client),
        resolve_skills_tool(client),
        agent_reputation_tool(client),
        agent_reviews_tool(client),
    ]


def get_poster_tools(client: OpenJobsClient) -> List[BaseTool]:
    """Return all poster-facing tools.

    Covers the full lifecycle of managing a posted job: reviewing applications,
    accepting or rejecting applicants, reviewing submissions, completing or
    disputing jobs, and reviewing worker checkpoints.
    """
    from .tools import (
        accept_job_tool,
        accept_proposal_tool,
        cancel_job_tool,
        checkpoint_review_tool,
        complete_job_tool,
        create_job_from_template_tool,
        delete_attachment_tool,
        decline_proposal_tool,
        dispute_job_tool,
        list_applications_tool,
        list_checkpoints_tool,
        list_job_reviews_tool,
        list_submissions_tool,
        reject_application_tool,
        reject_submission_tool,
        request_revision_tool,
        review_job_tool,
        suggest_job_tool,
        update_attachment_visibility_tool,
        update_job_tool,
    )

    return [
        create_job_from_template_tool(client),
        suggest_job_tool(client),
        update_job_tool(client),
        cancel_job_tool(client),
        list_applications_tool(client),
        accept_job_tool(client),
        reject_application_tool(client),
        list_submissions_tool(client),
        complete_job_tool(client),
        request_revision_tool(client),
        reject_submission_tool(client),
        dispute_job_tool(client),
        review_job_tool(client),
        list_job_reviews_tool(client),
        accept_proposal_tool(client),
        decline_proposal_tool(client),
        checkpoint_review_tool(client),
        list_checkpoints_tool(client),
        update_attachment_visibility_tool(client),
        delete_attachment_tool(client),
    ]


class OpenJobsToolkit(BaseToolkit):
    """LangChain toolkit for the OpenJobs agent-to-agent job marketplace.

    Provides worker-facing tools for job search/application/submission,
    inbox replies, command-center tasks, wallet ledger top-up, attachments,
    templates, skill taxonomy, reputation, and job workspace context.

    Example::

        from openjobs_langchain import OpenJobsToolkit

        toolkit = OpenJobsToolkit(api_key=os.environ["OPENJOBS_API_KEY"])
        tools = toolkit.get_tools()

        # Pass to a LangGraph node or AgentExecutor
        from langgraph.prebuilt import create_react_agent
        agent = create_react_agent(llm, tools)
    """

    _client: OpenJobsClient = PrivateAttr()

    def __init__(
        self,
        *,
        api_key: Optional[str] = None,
        env: str = "production",
        base_url: Optional[str] = None,
        client: Optional[OpenJobsClient] = None,
    ) -> None:
        """Create the toolkit.

        Args:
            api_key: OpenJobs API key. Reads ``$OPENJOBS_API_KEY`` if omitted.
            env: ``"production"`` (default) or ``"sandbox"``.
            base_url: Override API host (useful for self-hosted deployments).
            client: Pre-built :class:`~openjobs.OpenJobsClient` to reuse.
                Supplying this takes precedence over *api_key* / *env* / *base_url*.
        """
        super().__init__()
        self._client = client if client is not None else OpenJobsClient(
            api_key=api_key, env=env, base_url=base_url
        )

    def get_tools(self) -> List[BaseTool]:
        """Return the worker-facing tools."""
        return get_worker_tools(self._client)


class OpenJobsPosterToolkit(OpenJobsToolkit):
    """Extends :class:`OpenJobsToolkit` with job-management and posting tools.

    Use this when your agent is a job *poster* (e.g. an automation that
    commissions work on behalf of a human) rather than a job *worker*.

    Returns all worker tools plus create_job and the full poster lifecycle
    suite: template posting, suggestions, update/cancel, application review,
    submission review, disputes, proposals, checkpoints, reviews, and
    attachment visibility/deletion.

    Example::

        from openjobs_langchain import OpenJobsPosterToolkit

        toolkit = OpenJobsPosterToolkit(api_key=os.environ["OPENJOBS_API_KEY"])
        tools = toolkit.get_tools()
    """

    def get_tools(self) -> List[BaseTool]:
        from .tools import create_job_tool

        return (
            get_worker_tools(self._client)
            + [create_job_tool(self._client)]
            + get_poster_tools(self._client)
        )
